components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/logout/idpinit/processor/FederatedIdpInitLogoutProcessor.java

```diff
@@ -0,0 +1,632 @@
+
+            validateLogoutToken(signedJWT, identityProvider);
+            // Check for sid or sub claims in the logout token.
+            if (isSidClaimExists(claimsSet)) {
+                // Check whether the sid has an entry in the federated authentication session details table.
+                return logoutUsingSid((String) claimsSet.getClaim(OIDCAuthenticatorConstants.Claim.SID));
+            }
+            // Check whether the sub claim is available in the logout token.
+            if (log.isDebugEnabled()) {
+                log.debug("No sid presented in the logout token of the federated idp initiated logout request. Using " +
+                        "sub claim to terminate the sessions for tenant domain: " + tenantDomain);
+            }
+            // Retrieve the federated user id from the IDN_AUTH_USER table.
+            String sub;
+            String userId = null;
+            if (isSubClaimExists(claimsSet)) {
+                sub = claimsSet.getSubject();
+                userId = getUserId(tenantDomain, sub, identityProvider);
+                if (log.isDebugEnabled()) {
+                    log.debug("Trying OIDC federated identity provider initiated logout for the user: " + sub);
+                }
+            }
+            // Check whether the sub claim has a valid user in the IS.
+            return logoutUsingSub(userId);
+
+        } catch (ParseException e) {
+            throw handleLogoutClientException(ErrorMessages.LOGOUT_TOKEN_PARSING_FAILURE, e);
+        }
+    }
+
+    /**
+     * Terminate the session related to the sid value of the logout token.
+     *
+     * @param sid - sid claim included in the logout token.
+     * @return
+     * @throws LogoutServerException
+     */
+    private LogoutResponse.LogoutResponseBuilder logoutUsingSid(String sid)
+            throws LogoutServerException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Trying federated IdP initiated logout using sid: " + sid);
+        }
+        String sessionId = getSessionIdFromSid(sid);
+        if (StringUtils.isBlank(sessionId)) {
+            return new LogoutResponse.LogoutResponseBuilder(HttpServletResponse.SC_OK, StringUtils.EMPTY);
+        }
+        ServerSessionManagementService serverSessionManagementService =
+                OpenIDConnectAuthenticatorDataHolder.getInstance().getServerSessionManagementService();
+        if (serverSessionManagementService.removeSession(sessionId)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Session terminated for session Id: " + sessionId);
+            }
+            return new LogoutResponse.LogoutResponseBuilder(HttpServletResponse.SC_OK,
+                    OIDCAuthenticatorConstants.BackchannelLogout.LOGOUT_SUCCESS);
+        }
+        return new LogoutResponse.LogoutResponseBuilder(HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
```

Comment:
At which cases does serverSessionManagementService.removeSession(sessionId) return false?

Can we consider it as a SERVER error if it returns false.

Reply:
It return false only for empty sessionId. Since we already handle that, we can remove the if check.